query/src/main/scala/filodb/query/exec/MetadataExecPlan.scala

```diff
@@ -395,75 +348,96 @@ final case class LabelCardinalityExec(queryContext: QueryContext,
     s" startMs=$startMs, endMs=$endMs"
 }
 
-final case object TopkCardExec {
-  // TODO: tune this
-  // See here for help choosing a size:
-  //   https://datasketches.apache.org/docs/Frequency/FrequentItemsErrorTable.html
-  val MAX_ITEMSKETCH_MAP_SIZE = 256;
+/**
+ * Contains utilities for all TsCardinality materialize() derivatives.
+ */
+final case object TsCardExec {
+  val MAX_RESPONSE_SIZE = 5000  // TODO: tune this
+  val OVERFLOW_NAME = "_overflow_".utf8
+  val PREFIX_DELIM = ","  // not a utf8 in order to make use of mkString
+
+  case class CardCounts(active: Int, total: Int) {
+    if (total < active) {
+      qLogger.warn(s"CardCounts created with total < active; total: $total, active: $active")
+    }
+    def add(other: CardCounts): CardCounts = {
+      CardCounts(active + other.active,
+                 total + other.total)
+    }
+  }
+
+  /**
+   * Convenience class for converting between RowReader and TsCardExec data.
+   */
+  case class RowData(group: ZeroCopyUTF8String, counts: CardCounts) {
+    def toRowReader(): RowReader = {
+      SeqRowReader(Seq(group, counts.active, counts.total))
```

Comment:
This is ok, but you can consider creating a custom implementation of RowReader to avoid the Seq allocation. There are examples in code.

Reply:
Done!